### DISCUSSION TITLE
Add simulation methods to Thin Layers?

### DISCUSSION THREAD

**Discussion by JR-1991** [EnzymeML/PyEnzyme#37] [Ideas]
Now that we are able to perform kinetic modeling by utilizing COPASI and PySCeS, wouldn't it be a natural progression to add simulation?

Both libraries in essence provide simulation and the parameter estimation involves multiple steps of simulation. Thus, I think for the sake of exploration and testing from user-side it could be quite handy. The PySCeS Thin Layer already has a simulation method that iterates over all initial concentrations. Thus, it wouldn't be too much of work providing a method that simulates based on user paremetrization (init_concs)

If so, the method should be accessible at each Thin Layer, but for higher abstraction something like the following would be nice to have:

```python
enzmldoc.simulate("PySCeS", init_conc="My_Initial_Concs.yaml")
```

Where the `init_conc`-argument refers to either a YAML file with corresponding concentration or is a `Dict` instance that holds these values.

Open questions for this would be:

- Write the results back to the original document or return a new one? There is already as `isCalculated` tag for the `Replicate`class that could distinguish experimental from simulated data.
- Run COPASI and PySCeS simultaneously and get statistics? 
- Could be utilize any more frameworks that help parametrization such as PETab or SED-ML?

I think this would be an awesome addition that fits the scope of EnzymeML/PyEnzyme!

**Comment by fbergmann** [comment]
I can easily add a method that returns the time course simulations. But there is more information needed: 

* how long would you like to simulate for? (or should we look through the experiments and take the maximum time in there)
* how many data points would you like to receive (only at the time of measurements?,  equally spaced? ..)
* which model outputs would you like to have (only the ones also present in measurements? ...)
* do we need steady state too?

Also i would probably call the parameter `initial_values`, since you might want to reparameterize the whole model, not just the concentrations. 

I'm not sure what sort of statistics you would want from a simulation run. That would make sense for stochastic simulations, but i dont see how it applies to deterministic ones.
